pkg/storageincentives/proof.go

```diff
@@ -5,12 +5,197 @@
 package storageincentives
 
 import (
+	"errors"
+	"fmt"
+	"hash"
+	"math/big"
+
+	"github.com/ethersphere/bee/pkg/bmt"
 	"github.com/ethersphere/bee/pkg/bmtpool"
 	"github.com/ethersphere/bee/pkg/cac"
+	"github.com/ethersphere/bee/pkg/soc"
+	"github.com/ethersphere/bee/pkg/storageincentives/redistribution"
 	storer "github.com/ethersphere/bee/pkg/storer"
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
+var errMessage = errors.New("reserve commitment hasher: failure in proof creation")
+
+// returns the byte index of chunkdata where the spansize starts
+func spanOffset(sampleItem storer.SampleItem) uint8 {
+	ch := swarm.NewChunk(sampleItem.ChunkAddress, sampleItem.ChunkData)
+	if soc.Valid(ch) {
+		return swarm.HashSize + swarm.SocSignatureSize
+	}
+
+	return 0
+}
+
+// makeInclusionProofs creates transaction data for claim method.
+// In the document this logic, result data, is also called Proof of entitlement (POE).
+func makeInclusionProofs(
+	reserveSampleItems []storer.SampleItem,
+	anchor1 []byte,
+	anchor2 []byte,
+) (redistribution.ChunkInclusionProofs, error) {
+	if len(reserveSampleItems) != storer.SampleSize {
+		return redistribution.ChunkInclusionProofs{}, fmt.Errorf("reserve sample items should have %d elements", storer.SampleSize)
+	}
+	if len(anchor1) == 0 {
+		return redistribution.ChunkInclusionProofs{}, errors.New("anchor1 is not set")
+	}
+	if len(anchor2) == 0 {
+		return redistribution.ChunkInclusionProofs{}, errors.New("anchor2 is not set")
+	}
+
+	require3 := storer.SampleSize - 1
+	require1 := new(big.Int).Mod(new(big.Int).SetBytes(anchor2), big.NewInt(int64(require3))).Uint64()
+	require2 := new(big.Int).Mod(new(big.Int).SetBytes(anchor2), big.NewInt(int64(require3-1))).Uint64()
+	if require2 >= require1 {
+		require2++
+	}
+
+	// TODO: refactor, make it global / anchor (cleanup?)
+	prefixHasherFactory := func() hash.Hash {
+		return swarm.NewPrefixHasher(anchor1)
+	}
+	prefixHasherPool := bmt.NewPool(bmt.NewConf(prefixHasherFactory, swarm.BmtBranches, 8))
+
+	// Sample chunk proofs
+	rccontent := bmt.Prover{Hasher: bmtpool.Get()}
+	rccontent.SetHeaderInt64(swarm.HashSize * storer.SampleSize * 2)
+	rsc, err := sampleChunk(reserveSampleItems)
+	if err != nil {
+		return redistribution.ChunkInclusionProofs{}, errMessage
+	}
+	rscData := rsc.Data()
+	_, err = rccontent.Write(rscData[swarm.SpanSize:])
+	if err != nil {
+		return redistribution.ChunkInclusionProofs{}, errMessage
+	}
+	_, err = rccontent.Hash(nil)
+	if err != nil {
+		return redistribution.ChunkInclusionProofs{}, errMessage
+	}
+	proof1p1 := rccontent.Proof(int(require1) * 2)
```

Comment:
What is the theoretical maximum value for `require1`, can it always fit into the `int` range? The same goes for `require2`.